### PR TITLE
 Implement upgrade-aliases.sh to migrate a route-based k8s cluster to use IP aliases in GCE.

### PR DESCRIPTION
**Pull request by jingax10**
**What this PR does / why we need it**:
This PR implements upgrade-aliases.sh, which is used to migrate a route-based k8s cluster to use IP aliases in GCE.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
"NONE"
```

### CODE REVIEW

File: cluster/gce/upgrade-aliases.sh [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+# !!!EXPERIMENTAL!!! Upgrade a k8s cluster from routes to IP aliases for
+# node connectivity on GCE. This is only for migration.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+if [[ "${KUBERNETES_PROVIDER:-gce}" != "gce" ]]; then
+  echo "!!! ${1} only works on GCE" >&2
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: echo "Error: KUBERNETES_PROVIDER must be GCE"

File: cluster/gce/upgrade-aliases.sh [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+source "${KUBE_ROOT}/hack/lib/util.sh"
+source "${KUBE_ROOT}/cluster/kube-util.sh"
+
+# Print the number of routes used for k8s cluster node connectivity.
+#
+# Assumed vars:
+#   PROJECT
+function get-k8s-node-routes-count() {
+  local k8s_node_routes_count=$(gcloud alpha compute routes list \
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: Do we need to use the alpha API?

File: cluster/gce/upgrade-aliases.sh [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+# Detect the subnetwork where the k8s cluster resides.
+#
+# Assumed vars:
+#  KUBE_MASTER
+#  PROJECT
+#  ZONE
+# Vars set:
+#  IP_ALIAS_SUBNETWORK
+function detect-k8s-subnetwork() {
+  local subnetwork_url=$(gcloud alpha compute instances describe \
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: Can we use the `beta`?

File: cluster/gce/upgrade-aliases.sh [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    ${KUBE_MASTER} --project=${PROJECT} --zone=${ZONE} \
+    --format='value(networkInterfaces[0].subnetwork)')
+  if [ -n ${subnetwork_url} ]; then
+    IP_ALIAS_SUBNETWORK=$(echo ${subnetwork_url##*/})
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: What does `##*/` do? I'm not familiar with that

File: cluster/gce/upgrade-aliases.sh [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  local allow_subnet_cidr_routes_overlap
+  allow_subnet_cidr_routes_overlap=$(gcloud alpha compute networks subnets \
+    describe ${IP_ALIAS_SUBNETWORK} --project=${PROJECT} --region=${REGION} \
+    --format='value(allowSubnetCidrRoutesOverlap)')
+  if [ ${allow_subnet_cidr_routes_overlap,,} = $1 ]; then
+    echo "Subnet ${IP_ALIAS_SUBNETWORK}'s allowSubnetCidrRoutesOverlap is already set as $1"
+    return
+  fi
+
+  echo "Setting subnet \"${IP_ALIAS_SUBNETWORK}\" allowSubnetCidrRoutesOverlap to $1"
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: can you assign $1 to a variable?

`local allowOverlap=$1`

File: cluster/gce/upgrade-aliases.sh [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  local fingerprint=$(gcloud alpha compute networks subnets describe \
+    ${IP_ALIAS_SUBNETWORK} --project=${PROJECT} --region=${REGION} \
+    --format='value(fingerprint)')
+  local access_token=$(gcloud auth print-access-token)
+  local request="{\"allowSubnetCidrRoutesOverlap\":$1, \"fingerprint\":\"${fingerprint}\"}"
+  local subnetwork_url="${GCE_API_ENDPOINT}projects/${PROJECT}/regions/${REGION}/subnetworks/${IP_ALIAS_SUBNETWORK}"
+  until curl --header "Content-Type: application/json" --header "Authorization: Bearer ${access_token}" \
```
Comment (bowei) [inline cluster/gce/upgrade-aliases.sh]: Interesting, this is not supported via gcloud at the moment?

**Review by bowei — COMMENTED** [review]
Can you split off the go code change?

File: cluster/gce/upgrade-aliases.sh (hunk 1 above)
Comment (jingax10) [reply cluster/gce/upgrade-aliases.sh]: Done.

File: cluster/gce/upgrade-aliases.sh (hunk 2 above)
Comment (jingax10) [reply cluster/gce/upgrade-aliases.sh]: Not for getting route. Removed alpha from gcloud cmd.

**Comment by jingax10** [comment]
bowei@: the scoped go changes are in PR #56260.

**Comment by jingax10** [comment]
Reverted the codes (#56260). PTAL

**Comment by justinsb** [comment]
kops-aws e2e should be fixed by #57896

**Comment by justinsb** [comment]
/retest

quota issues in new aws account

**Comment by justinsb** [comment]
Opened https://github.com/kubernetes/kubernetes/issues/58578 to track the NFS failures.

As this has LGTM and the NFS failures are not deterministic, retesting...

/retest

**Comment by mikedanese** [comment]
@jingax10 Please cleanup history before merging. History should not contain merge commits or fixup commits.

**Comment by mikedanese** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#6-squashing-and-commit-titles

**Comment by Random-Liu** [comment]
@jingax10 @bowei ip-alias test is completely red after this change.

https://k8s-testgrid.appspot.com/google-gce#gci-gce-ip-alias

**Comment by jingax10** [comment]
Checking now ...

**Comment by jingax10** [comment]
I took a quick look. I found

`TEST_ETCD_IMAGE=2.2.1`
`TEST_ETCD_VERSION=2.2.1`

from [ci-kubernetes-e2e-gci-gce-ip-alias.env](https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gci-gce-ip-alias.env) matters. With those envs, I can reproduce the issue that no node can be registered to the cluster; without them, it works as expected.

I will take a further look to see my PR makes etcd version matters.

**Pull request merged**